File: Program 4/prog4CLF.py
# Collin L. Ferguson
# Collin.l.ferguson@ndus.edu
# Program 4 | written in Python | Working with: Files, Functions, Lists

import math


def main():
    fileFound = False
    outFile = None

    theList = []

    while fileFound == False:
        outFile = input("Please enter the name of the file: ")
        try:
            outFile = open(outFile, "r")
            fileFound = True
        except:
            print("That file could not be found.")

    for line in outFile:
        theList.append(int(line))

    print(findMaxValue(theList))
    print("-" * 40)
    print(findMinValue(theList))
    print("-" * 40)
    print(findFirstOccurrence(theList, 1))
    print(findFirstOccurrence(theList, 6))
    print("-" * 40)
    print(findLastOccurrence(theList, 1))
    print(findLastOccurrence(theList, 6))
    print("-" * 40)
    print(calculateAverage(theList))
    print("-" * 40)
    print(findNumberAboveAverage(theList))
    print("-" * 40)
    print(findNumberBelowAverage(theList))
    print("-" * 40)
    print("%1.6f" % standardDeviation(theList))


def findMaxValue(theList):
    intCurrentLargest = -2147483647

    for intCurrentValue in theList:
        if intCurrentValue > intCurrentLargest:
            intCurrentLargest = intCurrentValue
    return intCurrentLargest


def findMinValue(theList):
    intCurrentSmallest = 2147483647

    for intCurrentValue in theList:
        if intCurrentValue < intCurrentSmallest:
            intCurrentSmallest = intCurrentValue
    return intCurrentSmallest


def findFirstOccurrence(theList, intValueToFind):
    for intIndex in range(0, len(theList)):
        if theList[intIndex] == intValueToFind:
            return intIndex
    return -1


def findLastOccurrence(theList, intValueToFind):
    intLastOccurrence = -1
    for intIndex in range(0, len(theList)):
        if theList[intIndex] == intValueToFind:
            intLastOccurrence = intIndex
    return intLastOccurrence


def calculateAverage(theList):
    intAverage = 0
    for intIndex in range(0, len(theList)):
        intAverage += theList[intIndex]
    return intAverage / len(theList)


def findNumberAboveAverage(theList):
    fltAverage = calculateAverage(theList)
    intNumAboveAverage = 0

    for intIndex in range(0, len(theList)):
        if theList[intIndex] >= fltAverage:
            intNumAboveAverage += 1

    return intNumAboveAverage


def findNumberBelowAverage(theList):
    fltAverage = calculateAverage(theList)
    intNumAboveAverage = 0

    for intIndex in range(0, len(theList)):
        if theList[intIndex] < fltAverage:
            intNumAboveAverage += 1

    return intNumAboveAverage


def standardDeviation(theList):
    fltAverage = calculateAverage(theList)
    fltListAverage = 0

    for intIndex in range(0, len(theList)):
        fltListAverage += (theList[intIndex] - fltAverage)**2

    fltListAverage = math.sqrt(fltListAverage / len(theList))
    return fltListAverage


main()

File: Program 4/prog4CLF.c
/*
# Collin L. Ferguson
# Collin.l.ferguson@ndus.edu
# Program 4 | written in C | Working with: Files, Functions, Arrays
*/


#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void createSep()
{
    printf("\n-------------------------------\n");
}


int findMaxValue(int theArray[], int intArrayLength)
{
    int intCurrentLargest = INT_MIN;

    for (int intIndex = 0; intIndex <= intArrayLength -1; intIndex++)
    {
        if(theArray[intIndex] > intCurrentLargest)
        {
            intCurrentLargest = theArray[intIndex];
        }
    }

    return intCurrentLargest;
}


int findMinValue(int theArray[], int intArrayLength)
{
    int intCurrentSmallest = INT_MAX;

    for (int intIndex = 0; intIndex <= intArrayLength -1; intIndex++)
    {
        if(theArray[intIndex] < intCurrentSmallest)
        {
            intCurrentSmallest = theArray[intIndex];
        }
    }

    return intCurrentSmallest;
}


int findFirstOccurrence(int theArray[], int intArrayLength, int intValueToFind)
{
    for(int intIndex = 0; intIndex <= intArrayLength - 1; intIndex++)
    {
        if(theArray[intIndex] == intValueToFind)
        {
            return intIndex;
        }
    }
    return -1;
}


int findLastOccurrence(int theArray[], int intArrayLength, int intValueToFind)
{
    int intCurrentLastIndex = -1;
    for(int intIndex = 0; intIndex <= intArrayLength - 1; intIndex++)
    {
        if(theArray[intIndex] == intValueToFind)
        {
            intCurrentLastIndex = intIndex;
        }
    }
    return intCurrentLastIndex;
}


float calcualteAverage(int theArray[], int intArrayLength)
{
    float fltTotal = 0;
    for(int intIndex = 0; intIndex <= intArrayLength - 1; intIndex++)
    {
        fltTotal += theArray[intIndex];
    }
    return fltTotal / (float)intArrayLength;
}


int findNumbersAboveAverage(int theArray[], int intArrayLength)
{
    float fltAverage = calcualteAverage(theArray, intArrayLength);
    int intNumAboveAverage = 0;
    for (int intIndex = 0; intIndex <= intArrayLength - 1; intIndex++)
    {
        if(theArray[intIndex] >= fltAverage)
        {
            intNumAboveAverage += 1;
        }
    }
    return intNumAboveAverage;
}


int findNumbersBelowAverage(int theArray[], int intArrayLength)
{
    float fltAverage = calcualteAverage(theArray, intArrayLength);
    int intNumBelowAverage = 0;
    for(int intIndex = 0; intIndex <= intArrayLength - 1; intIndex++)
    {
        if(theArray[intIndex] < fltAverage)
        {
            intNumBelowAverage += 1;
        }
    }
    return intNumBelowAverage;
}


float standardDeviation(int theArray[], int intArrayLength)
{
    float fltAverage = calcualteAverage(theArray, intArrayLength);
    float fltListAverage = 0;

    for(int intIndex = 0; intIndex <= intArrayLength - 1; intIndex++)
    {
        fltListAverage += pow(theArray[intIndex] - fltAverage, 2);
    }

    fltListAverage = sqrt(fltListAverage / intArrayLength);
    return fltListAverage;

}


int main()
{
    int theArray[5];

    char strLine[10];
    char fileName[50];

    int bolFileFound = 0;

    FILE* inFile;

    while(bolFileFound == 0) //Reading from a file.
    {
        printf("Enter the name of the file: ");
        scanf("%[^\n]%*c", fileName);
        inFile = fopen(fileName, "r");

        if(inFile == NULL)
        {
            printf("Unable to open that file.\n");
        }
        else
        {
            bolFileFound = 1;
        }
    }

    int intIndex = 0;

    while(fgets(strLine, sizeof(strLine), inFile))
    {
        theArray[intIndex] = atoi(strLine);
        intIndex++;
    }

    fclose(inFile);

    int intArrayLength = intIndex; //sizeof(theArray) //static array size makes this impossible.

    printf("%d", findMaxValue(theArray, intArrayLength));
    createSep();
    printf("%d", findMinValue(theArray, intArrayLength));
    createSep();
    printf("%d", findMinValue(theArray, intArrayLength));
    createSep();
    printf("%d\n",findFirstOccurrence(theArray, intArrayLength, 6));
    printf("%d",findFirstOccurrence(theArray, intArrayLength, 1));
    createSep();
    printf("%d\n",findLastOccurrence(theArray, intArrayLength, 6));
    printf("%d",findLastOccurrence(theArray, intArrayLength, 1));
    createSep();
    printf("%1.6f",calcualteAverage(theArray, intArrayLength));
    createSep();
    printf("%d", findNumbersAboveAverage(theArray, intArrayLength));
    createSep();
    printf("%d", findNumbersBelowAverage(theArray, intArrayLength));
    createSep();
    printf("%1.6f",standardDeviation(theArray, intArrayLength));

    return 0;
}
